### PR TITLE
FIX: Fix interface and port configuration.

### LINKED ISSUE #222: Respect bind address passed to servers.
As noticed by @tacaswell, we currently ignore user input from Python and just use the configuration from the environment variables. The input from Python should take precedence (or at least not be silently ignored).

### PR DESCRIPTION
**Pull request by danielballan**
Closes #222. ~~More accurately, it _will_ once it's ready to merge -- right now this is just docs to help me sort out what to do.~~

### CODE REVIEW

**Comment by danielballan** [comment]
Do not merge. I'd like to take this opportunity to document what each of the environment variables does, which components pay attention to them and which Python parameters potentially override them.

File: caproto/server/common.py [hunk 1]
```diff
@@ -446,15 +449,15 @@ def __getitem__(self, pvname):
         addresses = get_beacon_address_list()
 
         while True:
-            if self.udp_sock is None:
+            if self.beacon_sock is None:
                 await self.async_layer.library.sleep(.1)
                 continue
             beacon = ca.RsrvIsUpResponse(13, self.port, self.beacon_count,
-                                         self.host)
+                                         self.interfaces[0])  # TODO which one?
```
Comment (klauer) [inline caproto/server/common.py]: Important TODO here - I think it should probably indicate the IP for each interface in the beacon message.

File: caproto/server/server.py [hunk 2]
```diff
@@ -968,7 +968,19 @@ def ioc_arg_parser(*, desc, default_prefix, argv=None, macros=None):
...
+    default_intf = get_server_address_list()
+    if default_intf == ['0.0.0.0']:
+        default_msg = '0.0.0.0'
+    else:
+        default_msg = (f"{' '.join(default_intf)} as specified by environment "
+                       f"variable EPICS_CAS_INTF_ADDR_LIST")
+    parser.add_argument('--interfaces', default=default_intf,
+                        nargs='+',
+                        help=(f"Interfaces to listen on. Default is "
+                              f"{default_msg}.  Multiple entires can be "
```
Comment (klauer) [inline caproto/server/server.py]: `entires`

Comment (klauer) [reply caproto/server/server.py]: Also, very nice having this in the help string.

File: caproto/tests/conftest.py [hunk 3]
```diff
@@ -456,25 +451,8 @@ def server(request):
...
-                        await ctx.run()
-                    except OSError as ex:
-                        # if we get an OS error, likely due to
-                        # port already is use, try again
-                        continue
-                    else:
-                        break
-                else:
-                    raise ex
+                ctx = caproto.curio.server.Context(pvdb)
```
Comment (klauer) [inline caproto/tests/conftest.py]: :+1:

File: caproto/trio/server.py [hunk 4]
```diff
@@ -111,27 +116,47 @@ def __init__(self, host, port, pvdb):
...
-                for addr, port in ca.get_server_address_list(self.port):
-                    self.log.debug('Listening on %s:%d', addr, port)
+                # TODO Should this interface be configurable?
+                await self.beacon_sock.bind(('0.0.0.0', 0))
+                # Find a random port number that is free on all interfaces,
+                # and get a bound TCP socket with that port number on each
+                # interface.
+                tcp_sockets = {}  # maps interface to bound socket
+                stashed_ex = None
+                for port in ca.random_ports(100):
```
Comment (klauer) [inline caproto/trio/server.py]: Is it possible to push this up in server common?

Comment (danielballan) [reply caproto/trio/server.py]: I started that way and pulled it apart when I realize that trio treats binding as asynchronous whereas curio and asyncio treat it as blocking. But, reflecting on your comment, I realized there are other places where only two of the three libs use code from common (and one is special-cased). So I moved the code in the base class for curio and asyncio and special-cased trio.

File: caproto/server/common.py (hunk 1 above)
Comment (danielballan) [reply caproto/server/common.py]: Does that mean we want a separate beacon loop (and `beacon_sock`) for each interface?

**Comment by danielballan** [comment]
I originally suggested waiting to merge this until thorough documentation could be included. I'd rather punt that to a separate issue.

This looks good to go but I'd feel better if @klauer could look over 4a5bda0.

**Pull request merged**